File: spark/src/test/scala/com/alectenharmsel/research/spark/MoabLicenseInfoTest.scala
package com.alectenharmsel.research.spark

import java.io.IOException
import java.util.ArrayList
import java.util.List
import org.apache.spark.rdd.RDD
import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.scalatest._

class MoabLicenseInfoTest extends FlatSpec with Matchers {

  val conf = new SparkConf().setMaster("local").setAppName("test")
  val sc = new SparkContext(conf)
  val data = sc.parallelize(
    Array[String](
      "05/11 22:58:25  MNodeUpdateResExpression(nyx5624,FALSE,TRUE)",
      "05/11 22:58:25  INFO:     License cfd_solv_ser        0 of   6 available  (Idle: 33.3%  Active: 66.67%)",
      "05/11 22:59:25  INFO:     License cfd_solv_ser        0 of   6 available  (Idle: 33.3%  Active: 66.67%)",
      "05/11 23:58:25  INFO:     License cfd_solv_ser        1 of   7 available  (Idle: 33.3%  Active: 66.67%)",
      "05/11 23:59:25  INFO:     License cfd_solv_ser        0 of   7 available  (Idle: 33.3%  Active: 66.67%)"
    )
  )
  val res = MoabLicenseInfo.run(data).collect()

  it should "return a single result" in {
    res.size should be (1)
  }

  it should "have 1 available" in {
    res(0)._3 should be (1)
  }

  it should "have 26 total" in {
    res(0)._4 should be (26)
  }

  it should "have a proper date" in {
    res(0)._1 should be ("05/11")
  }

  it should "have a proper license name" in {
    res(0)._2 should be ("cfd_solv_ser")
  }
}

File: hadoop/src/test/java/com/alectenharmsel/research/hadoop/MoabLicenseInfoTest.java
package com.alectenharmsel.research.hadoop;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mrunit.mapreduce.MapDriver;
import org.apache.hadoop.mrunit.mapreduce.ReduceDriver;
import org.apache.hadoop.mrunit.types.Pair;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

public class MoabLicenseInfoTest {

    private MapDriver mapDriver;
    private ReduceDriver reduceDriver;
    private ArrayList<Text> vals;
    private Text key;
    private List<Pair<Text, Text> > res;

    @Before
    public void setUp() {
        mapDriver = new MapDriver(new MoabLicenseInfo.Map());
        reduceDriver = new ReduceDriver(new MoabLicenseInfo.Reduce());

        key = new Text("cfd_solv_ser-05-11");
        vals = new ArrayList<Text>();
        vals.add(new Text("0,6"));
        vals.add(new Text("0,6"));
        vals.add(new Text("0,6"));
        vals.add(new Text("0,6"));
        vals.add(new Text("2,8"));
        vals.add(new Text("2,8"));
        vals.add(new Text("1,8"));
        vals.add(new Text("0,8"));
        reduceDriver.withInput(key, vals);
    }

    @Test
    public void testNoOutputNeeded() throws IOException {
        List<Pair<Text, Text> > res = mapDriver.withInput(new LongWritable(0),
            new Text(
                "05/11 22:58:25  MNodeUpdateResExpression(nyx5624,FALSE,TRUE)"
                )
            )
            .run();
        Assert.assertTrue(res.isEmpty());
    }

    @Test
    public void testLicenseLine() throws IOException {
        mapDriver.withInput(new LongWritable(0),
            new Text(
                "05/11 22:58:25  INFO:     License cfd_solv_ser        0 of   6 available  (Idle: 33.3%  Active: 66.67%)"
                )
            )
            .withOutput(new Text("cfd_solv_ser-05-11"), new Text("0,6"))
            .runTest();
    }

    @Test
    public void testResultSize() throws IOException {
        res = reduceDriver.run();
        Assert.assertEquals("Reducer should return a single result",
            1, res.size());
    }

    @Test
    public void testZeroAvailable() throws IOException {
        res = reduceDriver.run();
        String avg = res.get(0).getSecond().toString();
        double tmp = Double.parseDouble(avg.split(",")[0]);
        Assert.assertEquals("The average should be 5/8", 5.0/8.0, tmp, 0.01);
    }

    @Test
    public void testZeroTotal() throws IOException {
        res = reduceDriver.run();
        String total = res.get(0).getSecond().toString();
        double tmp = Double.parseDouble(total.split(",")[1]);
        Assert.assertEquals("The total should be 7.0", 7.0, tmp, 0.01);
    }

    @Test
    public void testZeroKey() throws IOException {
        res = reduceDriver.run();
        String key = res.get(0).getFirst().toString();
        Assert.assertEquals("The key should be properly formatted",
            "05-11,cfd_solv_ser", key);
    }
}
